sync/syncer.go

```diff
@@ -198,89 +216,191 @@ func (s *Syncer) getLayerFromNeighbors(currenSyncLayer types.LayerID) (*types.La
 	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
 }
 
-func (s *Syncer) fetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
-	output := s.fetchWithFactory(BlockReqFactory(blockIds), s.Concurrency)
+func (s *Syncer) FetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
+	output := s.fetchWithFactory(NewBlockWorker(s, s.Concurrency, BlockReqFactory(), blockSliceToChan(blockIds)))
 	blocksArr := make([]*types.Block, 0, len(blockIds))
 	for out := range output {
 		block := out.(*types.Block)
-
-		eligible, err := s.BlockEligible(&block.BlockHeader)
-		if err != nil {
-			s.Warning(fmt.Sprintf("failed checking eligiblety %v", block.ID()), err)
-			continue
-		}
-		if !eligible {
-			s.Warning(fmt.Sprintf("block %v not eligible", block.ID()), err)
-			continue
-		}
-
-		if err := s.syncMissingContent(block); err != nil {
-			s.Error(fmt.Sprintf("failed derefrencing block data %v", block.ID()), err)
+		if err := s.SyncAndValidate(block); err != nil {
+			s.Error(fmt.Sprintf("failed derefrencing block data %v %v", block.ID(), err))
 			continue
 		}
 
 		s.Info("added block to layer %v", block.Layer())
 		blocksArr = append(blocksArr, block)
-
 	}
 
 	return blocksArr, nil
 }
 
-func (s *Syncer) syncMissingContent(blk *types.Block) error {
+func (s *Syncer) SyncAndValidate(blk *types.Block) error {
+	s.Info("sync and validate block %v ", blk.ID())
+
+	//check if known
+	if _, err := s.GetBlock(blk.Id); err == nil {
+		s.Info("we already know this block %v ", blk.ID())
+		return nil
+	}
+
+	//check block signature and is identity active
+	if eligable, err := s.IsIdentityActive(blk.MinerID.Key, blk.Layer()); err != nil || !eligable {
+		return errors.New(fmt.Sprintf("block %v identity activation check failed ", blk.ID()))
+	}
+
+	//associated atx data availability
+	if blk.ATXID != *types.EmptyAtxId {
+		if _, err := s.syncAtxs([]types.AtxId{blk.ATXID}); err != nil {
+			return errors.New(fmt.Sprintf("cannot fetch associated atx for block %v %v ", blk.ID(), err))
+		}
+	}
+
+	//block eligibility
+	if eligable, err := s.BlockEligible(&blk.BlockHeader); err != nil || !eligable {
+		return errors.New(fmt.Sprintf("block %v eligablety check failed ", blk.ID()))
+	}
+
+	//data availability
+	txs, atxs, err := s.syncMissingContent(blk)
+	if err != nil {
+		return errors.New(fmt.Sprintf("data availabilty failed for block %v", blk.ID()))
+	}
+
+	//validate blocks view
+	if valid := s.ValidateView(blk); valid == false {
+		return errors.New(fmt.Sprintf("could not validate for block %v view ", blk.ID()))
+	}
+
+	if err := s.AddBlockWithTxs(blk, txs, atxs); err != nil {
```

Comment:
IMO this shouldn't be part of this method and should be the responsibility of the caller

Reply:
Also, you're passing the full list of ATXs (including already known ATXs and unknown ATXs). This causes you to re-insert the already known ATXs although it is not needed (there is a full flow behind it where you first try to fetch the atx from the db (although it is known) which is redundant

Reply:
i wanted the flow to be as simple as possible now also we have a cache so not that bad
but i agree ill find a better ay to do this